server/server/src/lib.rs

```diff
@@ -288,12 +295,67 @@ async fn run_server(
     Ok(restart)
 }
 
+pub fn logging_init(settings: Option<LoggingSettings>) {
+    let settings = settings.unwrap_or(LoggingSettings {
+        mode: LogMode::Console,
+        level: service::settings::Level::Info,
+        directory: None,
+        filename: None,
+        max_file_count: None,
+        max_file_size: None,
+    });
+    let config = match settings.mode {
+        LogMode::File => file_logger(&settings),
+        LogMode::Console => Config::new().console(),
+        LogMode::All => file_logger(&settings).console(),
+    };
+    fast_log::init(config.level(LevelFilter::from(settings.level.clone())))
+        .expect("Unable to initialise logger");
+}
+
+fn file_logger(settings: &LoggingSettings) -> fast_log::Config {
+    let default_log_file = "remote_server.log".to_string();
+    let default_log_dir = "log".to_string();
+    let default_max_file_count = 5;
+    let default_max_file_size = 10;
+
+    // Note: the file_split will panic if the path separator isn't appended
+    // and the path separator has to be unix-style, even on windows
+    let log_dir = &format!("{}/", settings.directory.clone().unwrap_or(default_log_dir),);
+    let log_path = &env::current_dir().unwrap_or_default().join(log_dir);
+    let log_file = settings
+        .filename
+        .clone()
+        .unwrap_or_else(|| default_log_file);
+    let log_file = match create_dir_all(log_path) {
+        Ok(_) => log_path.join(log_file).to_string_lossy().to_string(),
+        Err(_) => log_file.to_string(),
+    };
+    let max_file_count = settings.max_file_count.unwrap_or(default_max_file_count);
+    let max_file_size = settings.max_file_size.unwrap_or(default_max_file_size);
+
+    // file_loop will append to the specified log file until the max size is reached,
+    // then create a new log file with the same name, with date and time appended
+    // file_split will split the temp file when the max file size is reached
+    // and retain the max number of files while the server is running
+    // Note: when the server is started, the temp files are removed. The main log file is
+    // appended to, but only to the max size limit. Only one additional main log is created
+    Config::new()
+        .file_loop(&log_file, LogSize::MB(max_file_size))
+        .file_split(
```

Comment:
I did wonder if this whole section can be cleaned up a bit by using things like `path.parent()` in the `Path` struct.
Rather than separately creating directory path, file path, etc  but haven't thought through exactly how it would work or if it would actually be more confusing.
Want to share a couple of links just in case you weren't aware of it and/or were interested.
https://doc.rust-lang.org/stable/std/path/struct.Path.html
https://doc.rust-lang.org/stable/std/path/struct.PathBuf.html -> Could use this to combine directory and file names with OS Specific separators for example...

Given that fast_log requires linux style separators it's probably not going to help much I suspect.

Reply:
I see - file_split has this line:

```
        if !dir_path.is_empty() && !dir_path.ends_with("/") {
```

and yes, it also calls `create_dir_all`.
Will re-order the lines and try without calling it explicitly